op/model.py

```diff
@@ -231,26 +231,28 @@ def _load(self):
         return self._backend.config_get()
 
 
-class ResourceMapping(Mapping):
-    """Map of resource names to the local path for the resource."""
-    def __init__(self, resource_names, backend):
+class Resource:
+    """Object representing a resource for the charm.
+    """
+    def __init__(self, name, backend):
+        self.name = name
         self._backend = backend
-        self._data = {resource_name: None for resource_name in resource_names}
+        self.path = None
 
-    def __contains__(self, key):
-        return key in self._data
-
-    def __len__(self):
-        return len(self._data)
-
-    def __iter__(self):
-        return iter(self._data)
+    def fetch(self):
+        """Fetch the resource from the controller or store.
 
-    def __getitem__(self, resource_name):
-        resource_path = self._data[resource_name]
-        if resource_path is None:
-            resource_path = self._data[resource_name] = Path(self._backend.resource_get(resource_name))
-        return resource_path
+        If successfully fetched, this returns True and the resource file can be
+        accessed via the path attribute.
```

Comment:
This API will mean every single time you have a Resource you can't actually use it without checking if the path is valid. So the only benefit of having all resources conveniently accessible in a dict is for iteration, which seems like a minor use case compared to the more typical "I want that data!".

How about this instead:

    model.resources.fetch("name") => path

That'd cache things internally, and hold all the important logic in a very minimalist API. I'd _only_ expose this at first (no lists, no maps, etc), as I suspect this is what people will want in the vast majority of cases.  It's easy to grow an API carefully.. it's impossible to cut it down after the public release is out (at least until the next major, and even there it's painful).